### ISSUE TITLE
`Storage.resize_()` moves storage to current device

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#113300]
```
import torch

assert torch.cuda.device_count() >= 2

with torch.cuda.device(1):
    x = torch.zeros((2, 3), device="cuda:0")
    print(f"(before op) x.device: {x.device} x.untyped_storage().device: {x.untyped_storage().device}")
    x.untyped_storage().resize_(0)
    print(
        f"(after op) x.device: {x.device} x.untyped_storage().device: {x.untyped_storage().device}"
    )
```

```
(before op) x.device: cuda:0 x.untyped_storage().device: cuda:0
(after op) x.device: cuda:0 x.untyped_storage().device: cuda:1
```

<details>
<summary> Old Distributed Repro </summary>

```
torchrun --nproc_per_node=2 repro.py 
```

```
import os

import torch
from torch import distributed

local_rank = int(os.getenv("LOCAL_RANK", 0))
world_size = int(os.getenv("WORLD_SIZE", 1))
distributed.init_process_group(backend="nccl", rank=local_rank, world_size=world_size)
device = torch.device("cuda", local_rank)
print(f"rank={local_rank} world size={world_size} device={device} current device: {torch.cuda.current_device()}")

t = torch.randn((1024,), device=f"cuda:{local_rank}")
if local_rank == 1:
    print(f"[Rank {local_rank}] t.device: {t.device} t.untyped_storage().device: {t.untyped_storage().device}")
    print(f"[Rank {local_rank}] id(t.untyped_storage()): {id(t.untyped_storage())}")
    print(f"[Rank {local_rank}] t.untyped_storage().resize_(4).device: {t.untyped_storage().resize_(4).device} t.device: {t.device}")
    print(f"[Rank {local_rank}] id(t.untyped_storage()): {id(t.untyped_storage())}")
```

```
rank=1 world size=2 device=cuda:1 current device: 0
rank=0 world size=2 device=cuda:0 current device: 0
[Rank 1] t.device: cuda:1 t.untyped_storage().device: cuda:1
[Rank 1] id(t.untyped_storage()): 139630240323904
[Rank 1] t.untyped_storage().resize_(4).device: cuda:0 t.device: cuda:1
[Rank 1] id(t.untyped_storage()): 139630240323904
```
</details>

cc @ptrblck

**Comment by awgu** [comment]
It seems that the issue is that after the `resize_(0)` call, thereafter, the tensor's storage will use `cuda:0` on rank 1.

**Comment by gchanan** [comment]
This sounds fishy to me.  Can you check this on non-distributed native ops?  Meaning something like:
```
>>> with torch.cuda.device(1):
>>>    x = torch.zeros((2,3), device='cuda:0')
>>>    x.op()
``` 

where op {resize_, add_(1), [0]} 

My understanding is there are two cases:
1) most ops do consistent device checks, so x.op() should fail
2) there are some ops that don't do these that are probably special cased in the codegen somewhere; we should add these to the test set although I'm not sure how to find them.

**Comment by awgu** [comment]
1. `add_()`
```
with torch.cuda.device(1):
    x = torch.zeros((2, 3), device='cuda:0')
    x.add_(1)
```

```
(before op) x.device: cuda:0 x.untyped_storage().device: cuda:0
(after op) x.device: cuda:0 x.untyped_storage().device: cuda:0
```

2. `resize_(0)`
```
with torch.cuda.device(1):
    x = torch.zeros((2, 3), device="cuda:0")
    x.resize_(0)
```

```
(before op) x.device: cuda:0 x.untyped_storage().device: cuda:0
(after op) x.device: cuda:0 x.untyped_storage().device: cuda:0
```

3. `untyped_storage().resize_(0)`
```
with torch.cuda.device(1):
    x = torch.zeros((2, 3), device="cuda:0")
    x.untyped_storage().resize_(0)
```

```
(before op) x.device: cuda:0 x.untyped_storage().device: cuda:0
(after op) x.device: cuda:0 x.untyped_storage().device: cuda:1
```

**Comment by gchanan** [comment]
@awgu:
1) what if you use `storage()` instead of untyped storage?
2) what if you resize to a positive integer instead of 0?  (for both the `untyped_storage()` and `storage()` case)?

**Comment by awgu** [comment]
> 1. what if you use `storage()` instead of untyped storage?

- `storage().resize_(0)`: after, `untyped_storage().device` returns `cuda:1`; `storage().device` returns `cuda:1`

> 2. what if you resize to a positive integer instead of 0? (for both the untyped_storage() and storage() case)?

- `untyped_storage().resize_(2)`: after, `untyped_storage().device` returns `cuda:1`; `storage().device` returns `cuda:1`
- `storage().resize_(2)`: after, `untyped_storage().device` returns `cuda:1`; `storage().device` returns `cuda:1`

In short, they all still use current device (`cuda:1`).

**Comment by gchanan** [comment]
@ezyang @albanD is any of this surprising to you?

**Comment by albanD** [comment]
Definitely unexpected behavior that we should fix.
But a not so surprising oversight I would say? :p

**Comment by ezyang** [comment]
It shouldn't do this. We should fix it.

**Comment by ezyang** [comment]
cc @kurtamohler

**Comment by kurtamohler** [comment]
I can fix it
